h/static/scripts/features.js

```diff
@@ -21,58 +21,66 @@
  */
 'use strict';
 
-var retry = require('retry');
+var assign = require('core-js/modules/$.assign');
+
+var events = require('./events');
 
 var CACHE_TTL = 5 * 60 * 1000; // 5 minutes
 
-function features ($document, $http, $log) {
+// @ngInject
+function features ($document, $http, $log, $rootScope) {
   var cache = null;
-  var operation = null;
   var featuresUrl = new URL('/app/features', $document.prop('baseURI')).href;
+  var fetchOperation;
 
-  function fetch() {
-    // Short-circuit if a fetch is already in progress...
-    if (operation) {
-      return;
-    }
-    operation = retry.operation({retries: 10, randomize: true});
-
-    function success(data) {
-      cache = [Date.now(), data];
-      operation = null;
-    }
+  $rootScope.$on(events.USER_CHANGED, function () {
+    cache = null;
+  });
 
-    function failure(data, status) {
-      if (!operation.retry('failed to load - remote status was ' + status)) {
-        // All retries have failed, and we will now stop polling the endpoint.
-        $log.error('features service:', operation.mainError());
-      }
+  function fetch() {
+    if (fetchOperation) {
+      // fetch already in progress
+      return fetchOperation;
     }
 
-    operation.attempt(function () {
-      $http.get(featuresUrl)
-        .success(success)
-        .error(failure);
+    fetchOperation = $http.get(featuresUrl).then(function (response) {
+      cache = {
+        updated: Date.now(),
+        flags: response.data,
+      };
+    }).catch(function (err) {
+      // if for any reason fetching features fails, we behave as
+      // if all flags are turned off
+      $log.warn('failed to fetch feature data', err);
+      cache = assign({}, cache, {
+        updated: Date.now(),
+      });
+    }).then(function () {
+      fetchOperation = null;
```

Comment:
To recap a discussion on Slack, there isn't a race condition here because promise rejection/fulfillment handlers are always invoked on subsequent turns of the event loop.

I'm not too keen on promise.$resolved because it isn't part of the [$q API](https://code.angularjs.org/1.4.7/docs/api/ng/service/$q) or A+/ES6 promises.